dnsdist: Optionally keep all data in LRU cache

karelbilek opened the pull request:
### Short description

To be able to keep all data in cache, and not delete expired, we need another mechanism to evict from cache; and LRU is the most useful here, as the actually used elements will stay in the cache.

MaybeLruCache is either behaving like a regular map, if isLRU is false, or also has a linked list+map for constant-time putFront

One of the changes is also that finding in cache can now require a write lock.

I have hand-tested this code and it works for my purposes. I have not yet written any unit tests or regression tests.

For my local testing, I have had `assert` in the code to check that all the three containers have same size at all time; but I have removed it here.

### Checklist
I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] read and accepted the [Developer Certificate of Origin](https://github.com/PowerDNS/pdns/blob/master/DCO) document, including the [AI Policy](https://github.com/PowerDNS/pdns/blob/master/AI_POLICY.md), and added a ["Signed-off-by"](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md#developer-certificate-of-origin) to my commits
- [x] compiled this code
- [x] tested this code
- [ ] included documentation (including possible behaviour changes)
- [x] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

karelbilek commented:
SIEVE looks very easy to implement indeed. I will have a look at it

rgacogne commented:
There is only one thing that requires special attention: the `hand` iterator needs to be properly updated when the queue is modified.

rgacogne commented:
I think supporting only SIEVE is enough.

karelbilek commented on pdns/dnsdistdist/dnsdist-cache.hh:
```diff
@@ -32,6 +32,145 @@
 struct DNSQuestion;
 
+// lru cache is NOT locked; we need to use shared lock
+template <typename K, typename V>
```
As I am editing the code more, I start to wonder if the templating is worth it, when it's never being used anywhere else. I will keep it for now though

karelbilek committed:
dnsdist: Split packet cache get to getLocked

This is not yet doing any actual change; it is just to make the next commit diff smaller

Signed-off-by: Karel Bilek <kb@karelbilek.com>

karelbilek committed:
dnsdist: Implement SIEVE and LRU caches

Signed-off-by: Karel Bilek <kb@karelbilek.com>

rgacogne reviewed (commented):
I've only done a very high-level scan but I like it, thanks a lot!
The next step would be to test the new eviction algorithms in the regression tests as well, I think?
It's too late in the cycle for 2.1.0 but I'd like to have this in 2.2 :)

rgacogne commented:
I wonder if it would make sense to move to a `boost::multi_index_container` at this point. One hashed index would be the hash key, as it is today, one other sequenced index could be the SIEVE queue and we could add another sorted index on the TTL?

karelbilek commented:
I think I will get to it this/next week finally

On Thu, 26 Mar 2026 at 11:57, Karel Bílek ***@***.***> wrote:

> it's fine, I wanted to rework it for TTL buckets but unfortunately have
> other work
>
> On Thu 26. 3. 2026 at 11:56, Remi Gacogne ***@***.***>
> wrote:
>
>> *rgacogne* left a comment (PowerDNS/pdns#16692)
>> <https://github.com/PowerDNS/pdns/pull/16692?email_source=notifications&email_token=AAAZT4LWHXDB7QCAOD6N5WT4SUEFPA5CNFSNUABFM5UWIORPF5TWS5BNNB2WEL2JONZXKZKDN5WW2ZLOOQXTIMJTGM3DCMZXGU3KM4TFMFZW63VGMF2XI2DPOKSWK5TFNZ2LK4DSL5RW63LNMVXHIX3POBSW4X3DNRUWG2Y#issuecomment-4133613756>
>>
>> Sorry it's taking me so long to review this PR, I have been a bit swamped
>> with other things lately. I just opened #17039
>> <https://github.com/PowerDNS/pdns/pull/17039> that adds some benchmarks
>> for the cache, and I'll see how the new code performs compared to the
>> existing one shortly.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/PowerDNS/pdns/pull/16692?email_source=notifications&email_token=AAAZT4LWHXDB7QCAOD6N5WT4SUEFPA5CNFSNUABFM5UWIORPF5TWS5BNNB2WEL2JONZXKZKDN5WW2ZLOOQXTIMJTGM3DCMZXGU3KM4TFMFZW63VGMF2XI2DPOKSWK5TFNZ2LK4DSL5RW63LNMVXHIX3POBSW4X3DNRUWG2Y#issuecomment-4133613756>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAAZT4OH6AHCXEPBXKTRQ6L4SUEFPAVCNFSM6AAAAACQMFO242VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHM2DCMZTGYYTGNZVGY>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>
